src/incremental/serialize.ml

```diff
@@ -43,49 +40,78 @@ let results_exist () =
   let r_str = Fpath.to_string r in
   Sys.file_exists r_str && Sys.is_directory r_str
 
-(* Convenience enumeration of the different data types we store for incremental analysis, so file-name logic is concentrated in one place *)
-type incremental_data_kind = SolverData | CilFile | VersionData | AnalysisData
-
-let type_to_file_name = function
-  | SolverData -> solver_data_file_name
-  | CilFile -> cil_file_name
-  | VersionData -> version_map_filename
-  | AnalysisData -> analysis_data_file_name
-
-(** Used by the server mode to avoid serializing the solver state to the filesystem *)
-let server_solver_data : Obj.t option ref = ref None
-let server_analysis_data : Obj.t option ref = ref None
-
-(** Loads data for incremental runs from the appropriate file *)
-let load_data (data_type: incremental_data_kind) =
-  if server () then
-    match data_type with
-    | SolverData -> !server_solver_data |> Option.get |> Obj.obj
-    | AnalysisData -> !server_analysis_data |> Option.get |> Obj.obj
-    | _ -> failwith "Can only load solver and analysis data"
-  else
-    let p = Fpath.(gob_results_dir Load / type_to_file_name data_type) in
-    unmarshal p
-
-(** Stores data for future incremental runs at the appropriate file, given the data and what kind of data it is. *)
-let store_data (data : 'a) (data_type : incremental_data_kind) =
-  if server () then
-    match data_type with
-    | SolverData -> server_solver_data := Some (Obj.repr data)
-    | AnalysisData -> server_analysis_data := Some (Obj.repr data)
-    | _ -> ()
-  else (
+(** Module to cache the data for incremental analaysis during a run, before it is stored to disk, as well as the server mode *)
+module Cache = struct
+  type t = {
+    solver_data: Obj.t option ref;
+    analysis_data: Obj.t option ref;
+    version_data: MaxIdUtil.max_ids option ref;
+    cil_file: Cil.file option ref;
+  }
+
+  let data = ref {
+      solver_data = ref None ;
+      analysis_data = ref None ;
+      version_data = ref None ;
+      cil_file = ref None ;
+    }
+
+  (* GADT that may be used to query data from the cache *)
+  type _ data_query =
+    | SolverDataRequest : Obj.t data_query
+    | CilFileRequest : Cil.file data_query
+    | VersionDataRequest : MaxIdUtil.max_ids data_query
+    | AnalysisDataRequest : Obj.t data_query
+
+  (* Data type to pass incremental data into the cache *)
+  type incremental_data =
+    | SolverData of Obj.t
+    | CilFile of Cil.file
+    | VersionData of MaxIdUtil.max_ids
+    | AnalysisData of Obj.t
```

Comment:
This duplication seems unnecessary. I think just `data_query` would suffice and `update_data` could have a type like `type a. a data_query -> a -> unit`.